### PR TITLE
Lightmap editor (WIP)

### PR DESCRIPTION
**Pull request by 2020karaulov2020**
Face ligtmap editor.

### CODE REVIEW

**Comment by 2020karaulov2020** [comment]
@wootguy  
dEyQ3ar 1
1INfYxb 1

**Comment by wootguy** [comment]
Awesome!

I tried it out on de_dust and it works well on an 8x8 lightmap near T spawn. For most other lightmaps though it's pretty broken. The lightmap gets partially filled with random data which changes when I click `Reload`, and when I draw with a red color I get blue instead, and sometimes other colors.

image

It's ok if you don't want to fix that, but please add a "WIP" text somewhere explaining these bugs (like in the face editor).

**Comment by 2020karaulov2020** [comment]
@wootguy i see IMGUI bad works with images non two multiple ?

**Comment by 2020karaulov2020** [comment]
@wootguy okay I added WIP text. + some fixes for future

**Comment by 2020karaulov2020** [comment]
@wootguy issue at imgui github https://github.com/ocornut/imgui/issues/3829

**Comment by 2020karaulov2020** [comment]
@wootguy I can enable editing lightmap for all 4 styles, but i don't know about styles. I don't see changes in game if edit other styles.

You can simple uncomment here line to enable editing for all 4 styles : https://github.com/wootguy/bspguy/blob/d4a0061478955cb6a997c2f04eadf43bbbbad8b0/src/editor/Gui.cpp#L3128

File: src/gl/Texture.cpp [hunk 1]
```diff
@@ -38,10 +38,20 @@ void Texture::upload(int format)
...
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	if (lightmap)
+	{
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
```
Comment (wootguy) [inline src/gl/Texture.cpp:46]: easier to edit this way 👍

**Review by wootguy — CHANGES_REQUESTED** [review]
> i see IMGUI bad works with images non two multiple ?
yeah looks like it happens with NPOT textures. There might be something in the documentation about that.

**Comment by wootguy** [comment]
> @wootguy I can enable editing lightmap for all 4 styles, but i don't know about styles. I don't see changes in game if edit other styles.

The other styles are for toggled lights. I don't think there are many of those in CS. It would be hard to edit those now since you can't manually toggle lights on/off in the editor.

File: src/editor/Gui.cpp [hunk 2]
```diff
@@ -2904,6 +2912,369 @@ void Gui::drawEntityReport() {
...
+	static Texture* currentlightMap[MAXLIGHTMAPS] = { nullptr, };
+	static float colourPatch[3];
+
+	int readLightmaps = /*MAXLIGHTMAPS*/ 1;
+	int windowWidth = 50 + readLightmaps * 200;
+
+	ImGui::SetNextWindowSize(ImVec2(windowWidth, 520), ImGuiCond_FirstUseEver);
+	ImGui::SetNextWindowSizeConstraints(ImVec2(windowWidth, 520), ImVec2(windowWidth, 520));
+
+	if (ImGui::Begin("LightMap Editor (WIP)", &showLightmapEditorWidget)) {
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:3134]: Strange but it works even for 15x16, but not works for 11x15 11x16...) Possible WIDTH need power of two ))

File: src/editor/Gui.cpp [hunk 3]
```diff
@@ -2904,6 +2912,369 @@ void Gui::drawEntityReport() {
...
+		}
+		else
+		{
+			ImGui::Text("No selected face");
+			ImGui::Separator();
+
+			if (ImGui::Button("OK", ImVec2(120, 0)))
+			{
+
+			}
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:3272]: @wootguy I don't know how to close current window))

[the remark this answers is not in the document]
Comment (wootguy) [reply src/editor/Gui.cpp:3272]: to close set `showLightmapEditorWidget = false;`

I think it's better without the button though, and consistent with the other widgets.

**Comment by 2020karaulov2020** [comment]
> > @wootguy I can enable editing lightmap for all 4 styles, but i don't know about styles. I don't see changes in game if edit other styles.
> 
> The other styles are for toggled lights. I don't think there are many of those in CS. It would be hard to edit those now since you can't manually toggle lights on/off in the editor.

You know style names ? I can add it to lightmap editor)

**Comment by wootguy** [comment]
Lights can have styles with these names:
https://twhl.info/wiki/page/light

not sure that's what you meant though.

**Comment by 2020karaulov2020** [comment]
@wootguy you missing **glPixelStorei(GL_UNPACK_ALIGNMENT, 1);** in texture code for RGB format!

**Comment by wootguy** [comment]
Good find! Weird that it didn't show artifacts until now.

**Comment by 2020karaulov2020** [comment]
@wootguy all fixed and added lightmap edit for all styles))

But I found new problem ) Sometime crashes when do reloadLightmaps.

**Comment by 2020karaulov2020** [comment]
@wootguy now all works :)

**Comment by 2020karaulov2020** [comment]
1PB6Knq 1

**Comment by 2020karaulov2020** [comment]
@wootguy may be need add pixel alignment 1 byte for any texture format ?

File: src/editor/Gui.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp:3135]: Fixed)

File: src/gl/Texture.cpp (hunk 1 above)
Comment (2020karaulov2020) [reply src/gl/Texture.cpp:46]: I think yes

File: src/editor/Gui.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: Ok

**Comment by wootguy** [comment]
That fixed all the problems I saw earlier. Some comments on the window:
* The labels don't really make sense. All lightmaps could be considered "Light enabled". Maybe go with "Light 0", "Light 1", "Light 2", "Light 3".
* The lightmap size will be the same for all layers, so no need to show the size for all of them.
* would be nice to be able to resize the window to show the color picker without scrolling.

> may be need add pixel alignment 1 byte for any texture format ?

it's only needed for RGB textures. The default value of 4 is faster for RGBA.

File: src/editor/Gui.cpp [hunk 4]
```diff
@@ -2904,6 +2912,369 @@ void Gui::drawEntityReport() {
...
+					{
+						image_x = currentlightMap[i]->width;
+					}
+					if (image_y > currentlightMap[i]->height)
+					{
+						image_y = currentlightMap[i]->height;
+					}
+
+					int offset = (currentlightMap[i]->width * 3 * image_y) + (image_x * 3);
+					if (offset > currentlightMap[i]->width * currentlightMap[i]->height * 3)
```
[the remark this answers is not in the document]
Comment (2020karaulov2020) [reply src/editor/Gui.cpp]: added

**Comment by 2020karaulov2020** [comment]
@wootguy now all fixed ?))

Added turn off light in lightmap editor, and return back texture aligment for non RGB.

**Comment by 2020karaulov2020** [comment]
**All checked and should work now!**

**Comment by wootguy** [comment]
Nice that you you can toggle light styles now. I've got ideas for improvement, but it's working well enough to merge.

image

**Pull request merged**